### PR TITLE
Layer Norm Fusion

### PR DESCRIPTION
**Pull request by philipphack**
Rewrites layer norm patterns

(X, S, B, epsilon) -> (Y, [E, N]),

where, X, S and B are the input, scale and bias tensors, epsilon is a scalar and Y is the result of the normalization, into a Custom Call to the cuDNN library. Optionally also fuses the expectation E and the normalization factor N, which separately occur in the training graph, into the Custom Call.

### CODE REVIEW

**Comment by philipphack** [comment]
CC @reedwm, @nluehr.

**Review by reedwm — CHANGES_REQUESTED** [review]
Thanks for the PR!

Do you plan on adding the backwards pass of layer norm at some point?

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+}
+
+// Variance, expressed as E[(X - E[X])^2] or E[X^2] - E[X]^2 where E is the
+// expecation.
+auto Variance(HloInstruction** input0, HloInstruction** input1) {
+  return m::AnyOf<HloInstruction>(
+      Subtract(Expectation(Square(m::Op(input0))),
+               Square(Expectation(m::Op(input1)))),
+      Expectation(Square(Subtract(m::Op(input0), Expectation(m::Op(input1))))));
+}
```
[the remark this answers is not in the document]
Comment (tdanyluk) [reply xla/service/gpu/cudnn_norm_rewriter.cc:183]: @reedwm

TLDR: I think that currently the only good option is to compare the HLO pointers after the matching.

More thoughts (Sorry, this was interesting to me so I wrote a lot):

Let's say you want to match (A - A). The following kindof works:

(**Not recommended code:**)
```
auto SameOp(HloInstruction** existing_match) {
  return m::Op().WithPredicate(
      [existing_match](const HloInstruction* new_match) {
        return new_match == *existing_match;
      });
}

m::Subtract(m::Op(&lhs), SameOp(&lhs))
```

But this depends on the order of matching. For example if the Subtract pattern would match the operands in reverse order, it would stop working. Also, this reordering could very well happen with AnyOf, AnyOrder, etc. So I would not recommend doing this.

I think that to get a better solution we would have to implement a more complex matching algorithm which can detect whether a pattern is being matched the first time and behave differently based on that.
The tricky part is handling AnyOf, because if one of the branches fail, we have to forget all the matches that we found in that branch.

But if the Match function returned a map of matches instead of setting pointers, this could actually work.
The Match function's declaration could look like this:
```
std::pair<bool, PatternToHloInstructionMap> Match(PatternToHloInstructionMap matches_so_far);
```

I implemented this for NNEF Tools a few years ago, but in python:
https://github.com/KhronosGroup/NNEF-Tools/blob/5c928df05994ac3894512b5428513709bc72ad6d/nnef_tools/core/matcher.py#L306
Tensors can be matched multiple times with this.
The usage was something like this:
https://github.com/KhronosGroup/NNEF-Tools/blob/5c928df05994ac3894512b5428513709bc72ad6d/nnef_tools/optimization/data_format_optimizer.py#L321C1-L322C1

If you think that a new matcher library would be useful, feel free to create a bug, but whether we can do it depends on prioritization.

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 2]
```diff
@@ -6094,24 +6094,39 @@ GetCudnnFusedMHABackwardOperationGraph(
 }  // namespace
 
 static tsl::StatusOr<cudnn_frontend::ExecutionPlan> GetExecPlanFromHeuristics(
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/stream_executor/cuda/cuda_dnn.cc:6096]: The very small tensor sizes in the unit tests seem to be outside the design specification of some of the kernels in the library. I was told that the failure is expected behavior and that engine configs can at least in theory fail for other types of kernels as well.

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 3]
```diff
@@ -6094,24 +6094,39 @@ GetCudnnFusedMHABackwardOperationGraph(
...
+        cudnn_frontend::ExecutionPlanBuilder()
+            .setHandle(cudnn.handle())
+            .setEngineConfig(engine_config, opGraph.getTag())
+            .build();
+    if (plan.get_status() == CUDNN_STATUS_SUCCESS) {
+      return plan;
+    }
+  }
+
+  return tsl::errors::Internal("Failed to generate cuDNN execution plan.");
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/stream_executor/cuda/cuda_dnn.cc:6133]: Because this isn’t unexpected behavior, I only added a VLOG for now.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+#endif
+
+namespace xla {
+namespace gpu {
+
+namespace {
+
+namespace m = match;
+
+bool IsReduceAdd(const HloInstruction* instr) {
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: I aligned this more closely with AppliesMaxReduce in the convolution rewriter.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+auto MultiplyAnyOrder(HloInstruction** multiply, Pattern0 pattern0,
+                      Pattern1 pattern1) {
+  return OptionalConvertAndOrReshape(
+      m::MultiplyAnyOrder(multiply, pattern0, pattern1));
+}
+
+// Multiplication of pattern by itself with optional convert and/or reshape.
+template <typename Pattern>
+auto Square(Pattern pattern) {
+  return MultiplyAnyOrder(pattern, pattern)
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:141]: I think it’s easier to read with pattern repeated. I kept this for now.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+      m::Reduce(reduction, pattern, m::ConstantScalar(0))
+          .WithPredicate(
+              [](const HloInstruction* instr) { return IsReduceAdd(instr); }));
+}
+
+// Expected value, or mean, with optional broadcast.
+template <typename Pattern>
+auto Expectation(Pattern pattern) {
+  auto shared_subpattern =
+      MultiplyAnyOrder(m::Broadcast(m::ConstantScalar()), ReduceAdd(pattern));
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: It’s actually the reciprocal of the number of elements in the reduce dimensions. I added an approximate float comparison, which should be sufficient for this purpose.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+auto Expectation(HloInstruction** expectation, Pattern pattern) {
+  auto shared_subpattern = MultiplyAnyOrder(
+      expectation, m::Broadcast(m::ConstantScalar()), ReduceAdd(pattern));
+  return m::AnyOf<HloInstruction>(m::Broadcast(shared_subpattern),
+                                  shared_subpattern);
+}
+
+// Expected value, or mean, with optional broadcast.
+template <typename Pattern>
+auto Expectation(HloInstruction** expectation, HloInstruction** reduction,
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:189]: I don’t think it triggers undefined behavior as the pointer will still be owned by the caller at the time it is being modified. A conceptually similar alternative would be to always call the maximally capturing version and ignore the unnecessary captures. I didn’t particularly like the idea though. There are now slightly fewer overloads.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+        return inputx->unique_id() == input->unique_id() ||
+               inputx->operand_count() == 1 &&
+                   inputx->operand(0)->unique_id() == input->unique_id();
+      };
+      if (!is_input(input0) || !is_input(input1) || !is_input(input2)) {
+        VLOG(1) << "Layer norm operands not unique.";
+        return OkStatus();
+      }
+
+      // Skip initial convert, if present.
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:319]: That’s not the only reason for the existence of additional conversions. For instance, PAXML introduces a number of type conversions when certain flags are set.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+        VLOG(1)
+            << "Layer norm Custom Calls require Ampere or newer architecture.";
+        return OkStatus();
+      }
+
+      // Verify the uniqueness of the inputs.
+      auto is_input = [input](HloInstruction* inputx) -> bool {
+        return inputx->unique_id() == input->unique_id() ||
+               inputx->operand_count() == 1 &&
+                   inputx->operand(0)->unique_id() == input->unique_id();
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: This accounts for the occurrence of unary ops like reshapes and type converts. Direct comparison of pointers is functionally equivalent.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+      // Skip initial convert, if present.
+      if (input->opcode() == HloOpcode::kConvert) {
+        input = input->mutable_operand(0);
+      }
+
+      // Verify the element types. The types and shapes of the scale and bias
+      // must match.
+      if (!CompatibleElementType(input) || !CompatibleElementType(instr) ||
+          !CompatibleElementType(scale) || !CompatibleElementType(bias) ||
+          !ShapeUtil::Equal(scale->shape(), bias->shape())) {
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:338]: The current verification seems to be correct. Apparently, the public documentation will be updated.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+  }
+
+  // Matches and rewrites a layer norm pattern with scale multiplication and
+  // bias addition into a Custom Call to cuDNN. The pattern matching employs
+  // modularity to cover the scope of realizations.
+  Status MatchLayerNorm(HloInstruction* instr) {
+    HloInstruction *input, *input0, *input1, *input2, *epsilon, *scale, *bias,
+        *expectation, *reduce;
+    if (Match(instr,
+              SubtractMultiplyAddAnyOrder(
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:278]: Restricting the placement of reshapes and converts would limit the functionality of the rewriter. However, type conversions are now required to convert from and to BF16, FP16 and FP32 and reshapes may only modify degenerate dimensions.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+          custom_call->users()[0]->opcode() == HloOpcode::kGetTupleElement &&
+          custom_call->users()[0]->tuple_index() == 0) {
+        gte = custom_call->users()[0];
+      } else {
+        VLOG(1) << "Incompatible users of layer norm Custom Call.";
+        return OkStatus();
+      }
+
+      auto make_compatible_shape = [](Shape shape) -> Shape {
+        // Eliminate any leading degenerate dimensions.
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:542]: Scale and bias have the shape of the norm dimensions.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 13]
```diff
@@ -0,0 +1,629 @@
...
+  // bias addition into a Custom Call to cuDNN. The pattern matching employs
+  // modularity to cover the scope of realizations.
+  Status MatchLayerNorm(HloInstruction* instr) {
+    HloInstruction *input, *input0, *input1, *input2, *epsilon, *scale, *bias,
+        *expectation, *reduce;
+    if (Match(instr,
+              SubtractMultiplyAddAnyOrder(
+                  m::Op(&input),
+                  Expectation(&expectation, &reduce, m::Op(&input0)),
+                  NormFactor(&input1, &input2, &epsilon),
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: NormFactor now also captures the expectation.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+                                     int tuple_index) -> Status {
+        TF_ASSIGN_OR_RETURN(
+            HloInstruction * new_gte,
+            MakeGetTupleElementHlo(new_custom_call, tuple_index));
+        HloInstruction* new_instr = new_gte;
+        if (!ShapeUtil::Equal(new_gte->shape(), old_instr->shape())) {
+          TF_ASSIGN_OR_RETURN(new_instr,
+                              MakeReshapeHlo(old_instr->shape(), new_gte));
+        }
+        if (tuple_index != 2) {
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: The logic now evaluates the type of op that’s being replaced. I think that should make it easier to follow.

File: xla/service/gpu/cudnn_norm_rewriter.h [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+#define TENSORFLOW_COMPILER_XLA_SERVICE_GPU_CUDNN_NORM_REWRITER_H_
+
+#include "absl/strings/string_view.h"
+#include "xla/hlo/ir/hlo_module.h"
+#include "xla/service/hlo_pass_interface.h"
+
+namespace xla {
+namespace gpu {
+
+class CudnnNormRewriter : public HloModulePass {
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.h:27]: The idea is to also use this pass for other cases, including the backward pass of the layer norm and other types of norms. I kept the comment more general.

File: xla/service/gpu/cublas_cudnn.cc [hunk 16]
```diff
@@ -60,7 +60,7 @@ const absl::string_view kCudnnConvReorderFilterCallTarget =
 const absl::string_view kCudnnConvReorderFilterAndBiasCallTarget =
     "__cudnn$convReorderFilterAndBias";
 
-const absl::string_view kCubDeviceRadixSortTarget = "__cub$DeviceRadixSort";
+const absl::string_view kCudnnNormCallTarget = "__cudnn$norm";
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cublas_cudnn.cc:63]: This seemed to have been inserted between various cuDNN calls. Let me know if you want to revert this change.

File: xla/service/gpu/runtime/BUILD [hunk 17]
```diff
@@ -141,9 +141,23 @@ cc_library(
 )
 
 cc_library(
-    name = "fused_attention",
-    srcs = ["fused_attention.cc"],
-    hdrs = ["fused_attention.h"],
+    name = "conv_reorder",
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/runtime/BUILD:144]: It looked like this had been inserted between different types of convolutions.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 18]
```diff
@@ -0,0 +1,629 @@
...
+  return visitor.changed();
+}
+
+}  // anonymous namespace
+
+CudnnNormRewriter::CudnnNormRewriter(
+    se::CudaComputeCapability cuda_compute_capability)
+    : cuda_compute_capability_(cuda_compute_capability) {}
+
+StatusOr<bool> CudnnNormRewriter::Run(
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:638]: Do you know when Google will switch to 8.9.5?

File: xla/service/gpu/cudnn_norm_rewriter.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (reedwm) [reply xla/service/gpu/cudnn_norm_rewriter.h:27]: We should describe every type of pattern-match-and-replace that this pass does, since it's harder to determine this from the code. This should either be commented here in the header or on top of the function which does the rewrite, e.g. `MatchLayerNorm`. The current comment in `MatchLayerNorm` is good but it should also describe the definition of layer norm, especially considering how concise it is: `(x - Mean(x))/(Variance(x)**0.5) * scale + bias`

If you do comment in the function, you should add in the header comment: "For now, only the forward pass of layer norm is matched."

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (reedwm) [reply xla/stream_executor/cuda/cuda_dnn.cc:6133]: Isn't returning an Internal error here unexpected behavior?

File: xla/service/gpu/cudnn_norm_rewriter.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (reedwm) [reply xla/service/gpu/cudnn_norm_rewriter.cc:638]: > Do you know when Google will switch to 8.9.5?

I think in a few weeks. And we'll be able to manually test with 8.9.5 even sooner, likely by the time this PR is merged. So I'll be able to test performance.

File: xla/debug_options_flags.cc [hunk 19]
```diff
@@ -203,6 +203,7 @@ DebugOptions DefaultDebugOptionsIgnoringFlags() {
   opts.set_xla_gpu_ensure_minor_dot_contraction_dims(false);
   opts.set_xla_gpu_filter_kernels_spilling_registers_on_autotuning(true);
   opts.set_xla_gpu_llvm_verification_level(0);
+  opts.set_xla_gpu_enable_cudnn_layer_norm(false);
```
Comment (reedwm) [inline xla/debug_options_flags.cc:206]: Put this flag under `xla_gpu_enable_cudnn_fmha` and `xla_gpu_fused_attention_use_cudnn_rng`, since cudnn layer norm is a similar feature. And same in xla.proto.

File: xla/stream_executor/device_description.h [hunk 20]
```diff
@@ -68,6 +68,10 @@ struct CudaComputeCapability {
     return !(*this < CudaComputeCapability{other_major, other_minor});
   }
 
+  bool Is(int other_major, int other_minor = 0) const {
+    return !(*this == CudaComputeCapability{other_major, other_minor});
```
Comment (reedwm) [inline xla/stream_executor/device_description.h:72]: Why the `!`? Also don't think this function is necessary. In cudnn_norm_rewriter.cc, you should just check the major version is AMPERE or HOPPER.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 21]
```diff
@@ -0,0 +1,652 @@
...
+    HloInstruction *input, *input0, *input1, *input2, *scale, *bias, *epsilon,
+        *expectation, *expectation0, *reduce, *norm_factor, *variance;
+    if (Match(instr,
+              SubtractMultiplyAddAnyOrder(
+                  m::Op(&input),
+                  Expectation(&expectation, &reduce, m::Op(&input0)),
+                  NormFactor(&norm_factor, &input1, &input2, &variance,
+                             &expectation0, &epsilon),
+                  // NormFactor(&input1, &input2, &epsilon),
+                  m::Broadcast(m::Op(&scale)), m::Broadcast(m::Op(&bias))))) {
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:284]: Check that scale is broadcasted in the right dimension. This might be slightly tricky as the reshapes can change the effective normalized dimensions. Maybe add a test for the case where scale is broadcast in the wrong dimension.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_norm_rewriter.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdanyluk) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: I think that it's not good that the `HloInstruction* lhs, rhs;` are local to the SubtractFromSelf function, because the patterns will be matched after that function returns, so those variables will not be alive. Also the "instr" is unused.

> Would the WithPredicate be evaluated after lhs and rhs are assigned?
I think that it would. (WithPredicate defines an AllofPattern and that evaluates its parameters in order.)

(Honestly I would recommend just using the patterns in the existing way.)

[the remark this answers is not in the document]
Comment (reedwm) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: Hmm this might be possible by capturing an HloInstruction in the lambda. But for now, I'm fine with the PR's current approach.

@philipphack can you add a comment stating the caller should check that input0 is the same as input1?

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/stream_executor/cuda/cuda_dnn.cc:6133]: Individual plans can fail. All plans failing would be unexpected behavior.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 22]
```diff
@@ -325,6 +356,29 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
...
+                                .GetAsDouble({})
+                                .value();
+      int64_t numel = 1;
+      for (auto dim : norm_dims) {
+        numel *= reduce->operand(0)->shape().dimensions()[dim];
+      }
+      float rnumel = 1. / float(numel);
+      float numerical_epsilon = 128 * std::numeric_limits<bfloat16>::epsilon();
+      if (abs(actual_rnumel - rnumel) >
+          (abs(actual_rnumel + rnumel) * numerical_epsilon)) {
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc]: All these cases seem to be covered by the actual comparison.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 23]
```diff
@@ -254,43 +269,49 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
...
+        VLOG(1) << "Layer norm Custom Calls require Ampere or Hopper "
+                   "architectures.";
         return OkStatus();
       }
 
       // Verify the uniqueness of the inputs.
       auto is_input = [input](HloInstruction* inputx) -> bool {
         return inputx->unique_id() == input->unique_id() ||
-               inputx->operand_count() == 1 &&
+               inputx->opcode() == HloOpcode::kConvert &&
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:369]: I think we don’t need to verify the convert here since we verify later that input has a supported type. If the convert is to an unsupported type, then somewhere in the graph there would have to be a conversion from that unsupported type to a supported type.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 24]
```diff
@@ -0,0 +1,652 @@
...
+                   inputx->operand(0)->unique_id() == input->unique_id();
+      };
+      if (!is_input(input0) || !is_input(input1) || !is_input(input2) ||
+          expectation->unique_id() != expectation0->unique_id()) {
+        VLOG(1) << "Layer norm operands not unique.";
+        return OkStatus();
+      }
+
+      // Skip initial convert, if present.
+      if (input->opcode() == HloOpcode::kConvert) {
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:379]: Essentially the same as above. We currently verify that the operand of the convert has a supported type. If the result of the convert is not a supported type, then there would have to be a conversion from that type, or the entire graph would have to have that type.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 25]
```diff
@@ -0,0 +1,652 @@
...
+      TF_ASSIGN_OR_RETURN(
+          HloInstruction * reshape,
+          MakeReshapeHlo(reshaped_shape, transpose.value_or(input)));
+
+      // Reshape the scale and bias.
+      std::vector<int64_t> reshaped_scale_dims(reshaped_dims.begin() + 1,
+                                               reshaped_dims.end());
+      // cuDNN requires tensors to have at least four dimensions.
+      while (reshaped_scale_dims.size() < 4) {
+        reshaped_scale_dims.emplace_back(1);
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:486]: Shouldn't the shape of scale in your example be [3, 4, 5] instead of [1, 3, 4, 5]? We only augment it to [3, 4, 5, 1] because cuDNN requires inputs to have at least four dimensions.

[the remark this answers is not in the document]
Comment (reedwm) [reply xla/service/gpu/cudnn_norm_rewriter.cc:486]: > Shouldn't the shape of scale in your example be [3, 4, 5] instead of [1, 3, 4, 5]? We only augment it to [3, 4, 5, 1] because cuDNN requires inputs to have at least four dimensions.

The [table in the cudnn documentation for the NormalizationForward engine](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#normalizationforward) shows the scale/bias should have shape `[1,C,(D),H,W]`:

image

So the first dimension is always expected to be 1, according to the documentation.

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (reedwm) [reply xla/stream_executor/cuda/cuda_dnn.cc:6133]: > Individual plans can fail. All plans failing would be unexpected behavior.

My original comment was stating we should add a LOG of the first or last failing plan right before this Internal error. That way if an Internal error is raised, which is unexpected, the user can check the logs to see why at least one of the plans failed.

And ugh, GitHub does not always properly show the full comment thread when I respond to a comment. This comment here will probably appear as an individual comment when I post it, but as I'm typing it, it appears to be part of the comment chain

File: xla/debug_options_flags.cc (hunk 19 above)
Comment (reedwm) [reply xla/debug_options_flags.cc:206]: > Put this flag under `xla_gpu_enable_cudnn_fmha` and `xla_gpu_fused_attention_use_cudnn_rng`, since cudnn layer norm is a similar feature. And same in xla.proto.

You didn't move the flag in debug_options_flag.cc yet.

File: xla/stream_executor/device_description.h [hunk 26]
```diff
@@ -69,7 +69,7 @@ struct CudaComputeCapability {
   }
 
   bool Is(int other_major, int other_minor = 0) const {
-    return !(*this == CudaComputeCapability{other_major, other_minor});
+    return *this == CudaComputeCapability{other_major, other_minor};
   }
```
Comment (reedwm) [inline xla/stream_executor/device_description.h:73]: This function checks both the major and minor version. I see at https://developer.nvidia.com/cuda-gpus that some Ampere cards have compute capability 8.6 and so the check of `cuda_compute_capability.Is(se::CudaComputeCapability::AMPERE)` will not pass for these cards. I would remove this function and just check the major version in cudnn_norm_rewriter.cc.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 27]
```diff
@@ -170,33 +218,43 @@ auto AddReduce(HloInstruction** reduction, Pattern pattern) {
...
+          });
   return m::AnyOf<HloInstruction>(m::Broadcast(shared_subpattern),
                                   shared_subpattern);
 }
 
-// Variance, expressed as E[(X - E[X])^2] or E[X^2] - E[X]^2, where E is the
-// expectation.
+// Variance, expressed as expectation(X^2) - expectation(X)^2 or expectation((X
+// - expectation(X))^2).
 auto Variance(HloInstruction** expectation, HloInstruction** input0,
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:258]: Add a comment stating the caller should check input0 is the same as input1.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 28]
```diff
@@ -170,33 +218,43 @@ auto AddReduce(HloInstruction** reduction, Pattern pattern) {
 template <typename Pattern>
 auto Expectation(Pattern pattern) {
   auto shared_subpattern =
-      MultiplyAnyOrder(m::Broadcast(m::ConstantScalar()), AddReduce(pattern));
+      MultiplyAnyOrder(m::Broadcast(m::ConstantScalar()), AddReduce(pattern))
+          .WithPredicate([](const HloInstruction* instr) {
+            return CalculatesExpectation(instr);
+          });
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/cudnn_norm_rewriter.cc:225]: The broadcasted scalar is one over the number of elements in the reduced dimensions. We need the reduce instruction to verify the value.

**Comment by reedwm** [comment]
Is this ready for me to re-review?

**Comment by philipphack** [comment]
PTAL.

**Review by reedwm — APPROVED** [review]

**Comment by reedwm** [comment]
@philipphack can you merge/rebase main to fix merge conflicts?

**Review by reedwm — APPROVED** [review]
